src/reference/asciidoc/security.adoc

```diff
@@ -93,3 +102,108 @@ public class ContextConfiguration {
+
+Unfortunately propagation and population `SecurityContext` is just a half of work.
+Since the message isn't an owner of the threads in the message flow and we should be sure that we are secure against
+any incoming messages, we have to _clean up_ the `SecurityContext` from `ThreadLocal`, like it is done on the HTTP
+response commit.
+From other side we can't do that automatically in the Framework, so Spring Integration suggests another
+`ChannelInterceptor` for this purpose - `SecurityContextCleanupChannelInterceptor`.
+The `SecurityContext` cleanup (or restore to the previous state) is done from the `afterSendCompletion()` and
+`afterMessageHandled()` methods implementation.
+This interceptor can be applied to any message channel after the action on which you would like to free the current
+thread from the message's Principal.
+
+Consider we have a classical use-case when we'd like to split HTTP request to several pieces and aggregate in the end
+ before sending response.
+So, the config on the matter may look like this:
+
+[source,xml]
+----
+<int-http:inbound-gateway path="/foo" request-channel="splitChannel"/>
+
+<int:splitter input-channel="splitChannel" output-channel="processChannel"/>
+
+<int:channel id="processChannel">
+    <int:dispatcher task-executor="threadPollExecutor"/>
+</int:channel>
+
+<int:service-activator input-channel="processChannel" output-channel="aggregateChannel" ref="securedService"/>
+
+<int:aggregator input-channel="aggregateChannel"/>
+----
+
+The HTTP request can be secured with standard Spring Security mechanism.
+But we must be sure that `securedService` is invoked with the trusted Principal and since we are with that for the
+HTTP request, there is just enough to propagate the `SecurityContext` to the threads for that `processChannel`.
+For this purpose we use `SecurityContextPropagationChannelInterceptor` on that channel.
+But after the process is finished and pushed to the aggregator we leave the thread and it is returned to the pool to
+be ready for other invocations on the `processChannel`.
+In this case we must clean up `SecurityContext` manually using the `SecurityContextCleanupChannelInterceptor` on the
+same `processChannel`.
+
+For those cases when `MessageChannel` interception technique doesn't fit to the application logic the
+`SecurityContextCleanupAdvice` is presented, which can be configured as any other AOP `Advice` via standard Spring
+configuration or using `<request-handler-advice-chain>` on the Spring Integration component:
+
+[source,xml]
+----
+<int:channel id="queueChannel">
+    <int:queue/>
+    <int:interceptors>
+    	<bean class="org.springframework.integration.security.context.SecurityContextPropagationChannelInterceptor"/>
+    </int:interceptors>
+</int:channel>
+
+<int:outbound-channel-adapter channel="queueChannel" ref="securedService">
+    <int:request-handler-advice-chain>
+    	<bean class="org.springframework.integration.security.context.SecurityContextCleanupAdvice"/>
```

Comment:
This will clean up immediately. If there's an `output-channel` and they want to SECURE the whole flow, the advice has to go on the poller instead of in the RHAC.